README.md
# Dart-sudoku_solver_generator

A Dart library containing tools related to Sudoku puzzles. This includes generating, solving and handling puzzles.

> [![Latest_Release](https://img.shields.io/pub/v/sudoku_solver_generator)](https://pub.dev/packages/sudoku_solver_generator)
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)

## [Changelog](https://github.com/VarunS2002/Dart-sudoku_solver_generator/blob/main/sudoku_solver_generator/CHANGELOG.md)

## Usage

```yaml
dependencies:
  sudoku_solver_generator: ^2.0.0
```

A Sudoku Solver usage example:

```dart
import 'package:sudoku_solver_generator/sudoku_solver_generator.dart';

main() {
  var sudoku = <List<int?>>[
    [2, 1, 7, 4, 3, 5, 9, null, 8],
    [5, 8, 9, 0, 6, 2, 4, 3, null],
    [3, 4, 6, 0, 7, 9, 2, 5, 1],
    [1, 6, 2, 3, 0, 0, 5, 4, 0],
    [9, 0, 5, 2, 0, 4, 7, 8, 6],
    [4, 7, 8, 5, 9, 6, 1, 0, 0],
    [8, 0, 3, 0, 4, 1, 6, 7, 2],
    [0, 9, 4, 7, 2, 3, 8, 1, 5],
    [7, 2, 1, 0, 5, 0, 0, 0, 4],
  ];
  var sudokuSolved = SudokuSolver.solve(sudoku);
  SudokuUtilities.printSudoku(sudokuSolved);
}
```

A Sudoku Generator usage example:

```dart
import 'package:sudoku_solver_generator/sudoku_solver_generator.dart';

main() {
  var sudokuGenerator = SudokuGenerator(emptySquares: 54);
  SudokuUtilities.printSudoku(sudokuGenerator.newSudoku);
  print('');
  SudokuUtilities.printSudoku(sudokuGenerator.newSudokuSolved, animated: true);
}
```

<br>

> ### Sudoku game based on this package: [Flutter-Sudoku](https://github.com/VarunS2002/Flutter-Sudoku/)

## Features

- Generate Sudoku puzzles with your specified number of empty squares or clues.

- Option to specify if the required Sudoku should have a unique solution.

- Solve Sudoku puzzles of any difficulty instantly.

- Check if the Sudoku puzzle (unsolved or solved) is valid in terms of game rules.

- Check if the Sudoku puzzle has been solved properly.

- Check if the unsolved Sudoku has a unique solution.

- Convert the Data Structure of the Sudoku puzzle from a 2D List to 1D List and vice-versa.

- Print Sudoku puzzles to the console in an easy-to-read manner.

- Well documented client code.

## Note

- Referred Algorithms:

    - [Sudoku Generation Algorithm](https://www.geeksforgeeks.org/program-sudoku-generator/) in Java.

    - [Solving Algorithm](https://www.101computing.net/sudoku-generator-algorithm/) in Python.

    - [Sudoku Validation](https://www.geeksforgeeks.org/check-if-given-sudoku-board-configuration-is-valid-or-not/) in
      Python.

    - [Unique Solution Checking](https://github.com/ngbaanh/unique-solution-sudoku-generator) in C.

- If you face any issue or have suggestions then feel free to open an issue on GitHub.


sudoku_solver_generator/CHANGELOG.md
## 2.1.0+1 - [17-06-2021]

- Updated Documentation
    - Fixed usage example in `README.md`

## 2.1.0 - [17-06-2021]

- **BREAKING CHANGE**
    - Bool parameter `animated` has been removed from `SudokuUtilities.printSudoku()`
    - This is because this breaks support for exporting to web (dart2js)

- Added back support for dart2js

## 2.0.0 - [14-06-2021]

- **BREAKING CHANGE**
    - If you manually specified `emptySquares` for `SudokuGenerator()` then it will break as it is a named parameter now
    - If the manually specified `emptySquares` is more than 54 then it will break as it cannot generate a unique
      solution and will throw a `UnlikelyUniqueSolutionException`
    - Fix: `SudokuGenerator(emptySquares: 54)` instead of `SudokuGenerator(54)`
    - Fix: `SudokuGenerator(emptySquares: 60, uniqueSolution: false)` if `emptySquares` is more than 54

- Added the bool parameter `uniqueSolution` to `SudokuGenerator()` to specify if a puzzle should have a unique solution
    - Defaults to `true`
    - `UnlikelyUniqueSolutionException` is thrown if `emptySquares` is `> 54`
    - This is because a unique solution is unlikely if there are less than `27` clues
    - Fixes the known issue

- Added function `SudokuUtilities.hasUniqueSolution()` to check if a puzzle has a unique solution

- Improved function `SudokuUtilities.printSudoku()`
    - Improved the aesthetics of the output by adding spaces and new lines and removing commas and brackets
    - Added a bool parameter `animated` to animate it by printing the numbers one by one

- Improved documentation

## 1.1.0 - [08-06-2021]

- Improved checking if the Sudoku configuration is valid
    - `SudokuUtilities.isValidConfiguration()` now checks for invalid values in the puzzle
    - `{1, 2, 3, 4, 5, 6, 7, 8, 9, 0, null}` are considered valid

- Fixed RangeError in `SudokuUtilities.to2D()` if Sudoku dimensions aren't valid

## 1.0.0+2 - [06-06-2021]

- Updated Documentation
    - Description in `pubspec.yaml`

## 1.0.0+1 - [06-06-2021]

- Updated Documentation
    - Description in `README.md`
    - Usage in `README.md`

## 1.0.0 - [06-06-2021]

- First Release


sudoku_solver_generator/test/sudoku_solver_test.dart
import 'package:sudoku_solver_generator/sudoku_solver_generator.dart';
import 'package:test/test.dart';

void main() {
  group('SudokuSolver Tests', () {
    group('SudokuSolver solving tests', () {
      test('With null', () {
        var sudoku = <List<int?>>[
          [2, 1, 7, 4, 3, 5, 9, null, 8],
          [5, 8, 9, 0, 6, 2, 4, 3, null],
          [3, 4, 6, 0, 7, 9, 2, 5, 1],
          [1, 6, 2, 3, 0, 0, 5, 4, 0],
          [9, 0, 5, 2, 0, 4, 7, 8, 6],
          [4, 7, 8, 5, 9, 6, 1, 0, 0],
          [8, 0, 3, 0, 4, 1, 6, 7, 2],
          [0, 9, 4, 7, 2, 3, 8, 1, 5],
          [7, 2, 1, 0, 5, 0, 0, 0, 4],
        ];
        expect(SudokuUtilities.isSolved(SudokuSolver.solve(sudoku)), isTrue);
      });

      test('Null safe', () {
        var sudoku = <List<int>>[
          [2, 1, 7, 4, 3, 5, 9, 0, 8],
          [5, 8, 9, 0, 6, 2, 4, 3, 0],
          [3, 4, 6, 0, 7, 9, 2, 5, 1],
          [1, 6, 2, 3, 0, 0, 5, 4, 0],
          [9, 0, 5, 2, 0, 4, 7, 8, 6],
          [4, 7, 8, 5, 9, 6, 1, 0, 0],
          [8, 0, 3, 0, 4, 1, 6, 7, 2],
          [0, 9, 4, 7, 2, 3, 8, 1, 5],
          [7, 2, 1, 0, 5, 0, 0, 0, 4],
        ];
        expect(SudokuUtilities.isSolved(SudokuSolver.solve(sudoku)), isTrue);
      });

      test('Already Solved', () {
        var sudoku = <List<int>>[
          [2, 1, 7, 4, 3, 5, 9, 6, 8],
          [5, 8, 9, 1, 6, 2, 4, 3, 7],
          [3, 4, 6, 8, 7, 9, 2, 5, 1],
          [1, 6, 2, 3, 8, 7, 5, 4, 9],
          [9, 3, 5, 2, 1, 4, 7, 8, 6],
          [4, 7, 8, 5, 9, 6, 1, 2, 3],
          [8, 5, 3, 9, 4, 1, 6, 7, 2],
          [6, 9, 4, 7, 2, 3, 8, 1, 5],
          [7, 2, 1, 6, 5, 8, 3, 9, 4],
        ];
        expect(SudokuUtilities.isSolved(SudokuSolver.solve(sudoku)), isTrue);
      });

      test('Easy - 27', () {
        var sudoku = <List<int>>[
          [2, 7, 0, 1, 0, 3, 8, 0, 5],
          [0, 0, 9, 2, 0, 8, 3, 0, 7],
          [3, 5, 0, 7, 0, 6, 0, 1, 4],
          [4, 0, 7, 8, 0, 5, 9, 3, 0],
          [5, 6, 0, 4, 3, 0, 7, 8, 1],
          [9, 0, 3, 0, 0, 1, 5, 0, 0],
          [7, 3, 4, 0, 6, 2, 0, 5, 0],
          [8, 2, 5, 3, 0, 4, 6, 7, 0],
          [6, 9, 1, 0, 0, 7, 4, 2, 3],
        ];
        expect(SudokuUtilities.isSolved(SudokuSolver.solve(sudoku)), isTrue);
      });

      test('Hard - 54', () {
        var sudoku = <List<int>>[
          [0, 6, 0, 1, 0, 0, 0, 0, 0],
          [0, 8, 0, 0, 0, 0, 0, 0, 1],
          [0, 0, 0, 0, 9, 5, 0, 4, 6],
          [0, 0, 0, 0, 6, 1, 9, 7, 5],
          [0, 0, 6, 0, 5, 0, 4, 3, 0],
          [0, 0, 9, 7, 0, 0, 0, 0, 0],
          [0, 0, 0, 4, 1, 0, 5, 0, 0],
          [5, 0, 2, 0, 0, 3, 0, 0, 0],
          [0, 0, 0, 5, 0, 2, 0, 0, 0],
        ];
        expect(SudokuUtilities.isSolved(SudokuSolver.solve(sudoku)), isTrue);
      });

      test('Multiple solutions - 80', () {
        var sudoku = <List<int>>[
          [0, 0, 0, 0, 0, 0, 0, 0, 0],
          [0, 0, 0, 0, 0, 0, 0, 0, 0],
          [0, 0, 0, 0, 9, 0, 0, 0, 0],
          [0, 0, 0, 0, 0, 0, 0, 0, 0],
          [0, 0, 0, 0, 0, 0, 0, 0, 0],
          [0, 0, 0, 0, 0, 0, 0, 0, 0],
          [0, 0, 0, 0, 0, 0, 0, 0, 0],
          [0, 0, 0, 0, 0, 0, 0, 0, 0],
          [0, 0, 0, 0, 0, 0, 0, 0, 0],
        ];
        expect(SudokuUtilities.isSolved(SudokuSolver.solve(sudoku)), isTrue);
      });
    });

    group('SudokuSolver exception tests', () {
      test('Invalid Sudoku - Repetitive numbers', () {
        var sudoku = <List<int>>[
          [2, 1, 7, 4, 3, 5, 9, 2, 8],
          [5, 8, 9, 0, 6, 2, 4, 3, 0],
          [3, 4, 6, 0, 7, 9, 2, 5, 1],
          [1, 6, 2, 3, 0, 0, 5, 4, 0],
          [9, 0, 5, 2, 0, 4, 7, 8, 6],
          [4, 7, 8, 5, 9, 6, 1, 0, 0],
          [8, 0, 3, 0, 4, 1, 6, 7, 2],
          [0, 9, 4, 7, 2, 3, 8, 1, 5],
          [7, 2, 1, 0, 5, 0, 0, 0, 4],
        ];
        expect(() => SudokuSolver.solve(sudoku),
            throwsA(const TypeMatcher<InvalidSudokuConfigurationException>()));
      });

      test('Invalid Sudoku - Wrong numbers', () {
        var sudoku = <List<int>>[
          [2, 1, 7, 4, 3, 5, 9, 29, 8],
          [5, 8, 9, 0, 6, 2, 4, 3, 0],
          [3, 4, 6, 0, 7, 9, 2, 5, 1],
          [1, 6, 2, 3, 0, 0, 5, 4, 0],
          [9, 0, 5, 2, 0, 4, 7, 8, 6],
          [4, 7, 8, 5, 9, 6, 1, 0, 0],
          [8, 0, 3, 0, 4, 1, 6, 7, 2],
          [0, 9, 4, 7, 2, 3, 8, 1, 5],
          [7, 2, 1, 0, 5, 0, 0, 0, 4],
        ];
        expect(() => SudokuSolver.solve(sudoku),
            throwsA(const TypeMatcher<InvalidSudokuConfigurationException>()));
      });

      test('Invalid Sudoku - Wrong List dimensions (rows)', () {
        var sudoku = <List<int>>[
          [2, 1, 7, 4, 3, 5, 9, 6, 8],
          [5, 8, 9, 0, 6, 2, 4, 3, 0],
          [3, 4, 6, 0, 7, 9, 2, 5, 1],
          [1, 6, 2, 3, 0, 0, 5, 4, 0],
          [9, 0, 5, 2, 0, 4, 7, 8, 6],
          [4, 7, 8, 5, 9, 6, 1, 0, 0],
          [8, 0, 3, 0, 4, 1, 6, 7, 2],
          [0, 9, 4, 7, 2, 3, 8, 1, 5],
        ];
        expect(() => SudokuSolver.solve(sudoku),
            throwsA(const TypeMatcher<InvalidSudokuConfigurationException>()));
      });

      test('Invalid Sudoku - Wrong List dimensions (columns)', () {
        var sudoku = <List<int>>[
          [2, 1, 7, 4, 3, 5, 9, 6, 8],
          [5, 8, 9, 0, 6, 2, 4, 3, 0],
          [3, 4, 6, 0, 7, 9, 2, 5, 1],
          [1, 6, 2, 3, 0, 0, 5, 4],
          [9, 0, 5, 2, 0, 4, 7, 8, 6],
          [4, 7, 8, 5, 9, 6, 1, 0, 0],
          [8, 0, 3, 0, 4, 1, 6, 7, 2],
          [0, 9, 4, 7, 2, 3, 8, 1, 5],
          [7, 2, 1, 0, 5, 0, 0, 0, 4],
        ];
        expect(() => SudokuSolver.solve(sudoku),
            throwsA(const TypeMatcher<InvalidSudokuConfigurationException>()));
      });
    });
  });
}


sudoku_solver_generator/test/sudoku_generator_test.dart
import 'package:sudoku_solver_generator/sudoku_solver_generator.dart';
import 'package:test/test.dart';

void main() {
  group('SudokuGenerator Tests', () {
    group('SudokuGenerator unsolved tests', () {
      test('Beginner - 18', () {
        var generator = SudokuGenerator(emptySquares: 18);
        expect(
            SudokuUtilities.isValidConfiguration(generator.newSudoku), isTrue);
      });

      test('Easy - 27 (default)', () {
        var generator = SudokuGenerator();
        expect(
            SudokuUtilities.isValidConfiguration(generator.newSudoku), isTrue);
      });

      test('Medium - 36', () {
        var generator = SudokuGenerator(emptySquares: 36);
        expect(
            SudokuUtilities.isValidConfiguration(generator.newSudoku), isTrue);
      });

      test('Hard - 54', () {
        var generator = SudokuGenerator(emptySquares: 54);
        expect(
            SudokuUtilities.isValidConfiguration(generator.newSudoku), isTrue);
      });

      test('Unrealistic - 80', () {
        var generator =
            SudokuGenerator(emptySquares: 80, uniqueSolution: false);
        expect(
            SudokuUtilities.isValidConfiguration(generator.newSudoku), isTrue);
      });
    });

    group('SudokuGenerator unique solution tests', () {
      test('Beginner - 18', () {
        var generator = SudokuGenerator(emptySquares: 18);
        expect(SudokuUtilities.hasUniqueSolution(generator.newSudoku), isTrue);
      });

      test('Easy - 27 (default)', () {
        var generator = SudokuGenerator();
        expect(SudokuUtilities.hasUniqueSolution(generator.newSudoku), isTrue);
      });

      test('Medium - 36', () {
        var generator = SudokuGenerator(emptySquares: 36);
        expect(SudokuUtilities.hasUniqueSolution(generator.newSudoku), isTrue);
      });

      test('Hard - 54', () {
        var generator = SudokuGenerator(emptySquares: 54);
        expect(SudokuUtilities.hasUniqueSolution(generator.newSudoku), isTrue);
      });
    });

    group('SudokuGenerator solved tests', () {
      test('Beginner - 18', () {
        var generator = SudokuGenerator(emptySquares: 18);
        expect(SudokuUtilities.isValidConfiguration(generator.newSudokuSolved),
            isTrue);
      });

      test('Easy - 27 (default)', () {
        var generator = SudokuGenerator();
        expect(SudokuUtilities.isValidConfiguration(generator.newSudokuSolved),
            isTrue);
      });

      test('Medium - 36', () {
        var generator = SudokuGenerator(emptySquares: 36);
        expect(SudokuUtilities.isValidConfiguration(generator.newSudokuSolved),
            isTrue);
      });

      test('Hard - 54', () {
        var generator = SudokuGenerator(emptySquares: 54);
        expect(SudokuUtilities.isValidConfiguration(generator.newSudokuSolved),
            isTrue);
      });
    });

    group('SudokuGenerator exception tests', () {
      test('<1 emptySquares', () {
        expect(() => SudokuGenerator(emptySquares: 0),
            throwsA(const TypeMatcher<InvalidEmptySquaresException>()));
      });

      test('>81 emptySquares (no uniqueSolution)', () {
        expect(() => SudokuGenerator(emptySquares: 82, uniqueSolution: false),
            throwsA(const TypeMatcher<InvalidEmptySquaresException>()));
      });

      test('>54 emptySquares (uniqueSolution)', () {
        expect(() => SudokuGenerator(emptySquares: 55),
            throwsA(const TypeMatcher<UnlikelyUniqueSolutionException>()));
      });
    });
  });
}


sudoku_solver_generator/lib/src/sudoku_solver_base.dart
import 'package:sudoku_solver_generator/src/sudoku_utilities_base.dart';

/// Provides the functionality of solving Sudoku puzzles.
///
/// This class shouldn't be instantiated.
class SudokuSolver {
  /// Solves the provided [sudoku] and returns the solved puzzle.
  ///
  /// Might return different solutions everytime if the puzzle has more than
  /// one solution.
  /// The [sudoku] must be a [List] of 9 [List]s of 9 [int]s and empty squares
  /// should be represented by `null` or `0`.
  ///
  /// [InvalidSudokuConfigurationException] is thrown if the configuration of
  /// the [sudoku] is not valid.
  static List<List<int>> solve(List<List<int?>> sudoku) {
    var _solvedGrid = SudokuUtilities.makeNullSafe(sudoku);
    _solveImplementation(_solvedGrid);
    return _solvedGrid;
  }

  static List<int>? _findEmpty(List<List<int>> sudoku) {
    for (var i = 0; i < 9; i++) {
      for (var j = 0; j < 9; j++) {
        if (sudoku[i][j] == 0) {
          return [i, j];
        }
      }
    }
    return null;
  }

  static bool _valid(List<List<int>> sudoku, int number, List<int> position) {
    for (var i = 0; i < 9; i++) {
      if (sudoku[position[0]][i] == number && position[1] != i) {
        return false;
      }
    }
    for (var i = 0; i < 9; i++) {
      if (sudoku[i][position[1]] == number && position[0] != i) {
        return false;
      }
    }
    var boxX = position[1] ~/ 3;
    var boxY = position[0] ~/ 3;
    for (var i = boxY * 3; i < boxY * 3 + 3; i++) {
      for (var j = boxX * 3; j < boxX * 3 + 3; j++) {
        if (sudoku[i][j] == number && [i, j] != position) {
          return false;
        }
      }
    }
    return true;
  }

  static bool _solveImplementation(List<List<int>> sudoku) {
    int row;
    int column;
    if (_findEmpty(sudoku) == null) {
      return true;
    } else {
      row = _findEmpty(sudoku)![0];
      column = _findEmpty(sudoku)![1];
    }
    for (var i = 1; i < 10; i++) {
      if (_valid(sudoku, i, [row, column])) {
        sudoku[row][column] = i;
        if (_solveImplementation(sudoku)) {
          return true;
        }
        sudoku[row][column] = 0;
      }
    }
    return false;
  }
}


sudoku_solver_generator/lib/src/sudoku_generator_base.dart
import 'package:sudoku_solver_generator/src/sudoku_exceptions_base.dart';
import 'package:sudoku_solver_generator/src/sudoku_utilities_base.dart';

/// Provides the functionality of generating Sudoku puzzles.
///
/// This class has to be instantiated for every new puzzle.
class SudokuGenerator {
  /// Constructor to generate the Sudoku. Use the getter methods to retrieve
  /// the puzzle.
  ///
  /// [emptySquares] specifies how many boxes should be left empty in the
  /// unsolved puzzle. Clues will be 81-[emptySquares].
  /// Defaults to `27`.
  /// Higher difficulty is indicated by more [emptySquares].
  ///
  /// Suggested values for various difficulty levels:
  /// * Beginner - 18
  /// * Easy - 27
  /// * Medium - 36
  /// * Hard - 54
  ///
  /// [uniqueSolution] specifies if the required Sudoku must have
  /// only 1 solution.
  /// Defaults to `true`.
  ///
  ///
  /// [InvalidEmptySquaresException] is thrown if the value of [emptySquares]
  /// is invalid.
  /// Valid values: Any integer from 1-81 (inclusive).
  ///
  /// [UnlikelyUniqueSolutionException] is thrown if [emptySquares] is `> 54`
  /// when [uniqueSolution] is `true`.
  /// This is because a unique solution is unlikely if there are less than
  /// `27` clues.
  SudokuGenerator({int emptySquares = 27, bool uniqueSolution = true}) {
    if (emptySquares < 1) {
      throw InvalidEmptySquaresException();
    }
    if (uniqueSolution) {
      if (emptySquares > 54) {
        throw UnlikelyUniqueSolutionException();
      }
    } else {
      if (emptySquares > 81) {
        throw InvalidEmptySquaresException();
      }
    }
    _uniqueSolution = uniqueSolution;
    _emptySquares = emptySquares;
    _sudoku = List.generate(9, (i) => List.generate(9, (j) => 0));
    _fillValues();
  }

  late bool _uniqueSolution;
  late int _emptySquares;
  late List<List<int>> _sudoku;
  late List<List<int>> _sudokuSolved;

  /// Unsolved Sudoku with the specified [emptySquares].
  List<List<int>> get newSudoku => _sudoku;

  /// 1 possible solution of the unsolved Sudoku if [uniqueSolution] is `false`
  /// else the only solution.
  ///
  /// When [uniqueSolution] is `false` use [SudokuUtilities.isSolved()] to
  /// check if a game is solved instead of comparing with this.
  /// This is to prevent false-negatives when a puzzle has been solved
  /// correctly but the solution is not identical to this one.
  List<List<int>> get newSudokuSolved => _sudokuSolved;

  void _fillValues() {
    _fillDiagonal();
    _fillRemaining(0, 3);
    _sudokuSolved = SudokuUtilities.copySudoku(_sudoku);
    _removeClues();
  }

  void _fillDiagonal() {
    for (var i = 0; i < 9; i = i + 3) {
      _fillBox(i, i);
    }
  }

  void _fillBox(int row, int column) {
    int number;
    for (var i = 0; i < 3; i++) {
      for (var j = 0; j < 3; j++) {
        do {
          number = _randomGenerator();
        } while (!_unUsedInBox(row, column, number));
        _sudoku[row + i][column + j] = number;
      }
    }
  }

  bool _unUsedInBox(int rowStart, int columnStart, int number) {
    for (var i = 0; i < 3; i++) {
      for (var j = 0; j < 3; j++) {
        if (_sudoku[rowStart + i][columnStart + j] == number) {
          return false;
        }
      }
    }
    return true;
  }

  static int _randomGenerator() {
    var numberList = <int>[1, 2, 3, 4, 5, 6, 7, 8, 9]..shuffle();
    return numberList[0];
  }

  bool _checkIfSafe(int i, int j, int number) =>
      (_unUsedInRow(i, number) && _unUsedInColumn(j, number)) &&
      _unUsedInBox(i - (i % 3), j - (j % 3), number);

  bool _unUsedInRow(int i, int number) {
    for (var j = 0; j < 9; j++) {
      if (_sudoku[i][j] == number) {
        return false;
      }
    }
    return true;
  }

  bool _unUsedInColumn(int j, int number) {
    for (var i = 0; i < 9; i++) {
      if (_sudoku[i][j] == number) {
        return false;
      }
    }
    return true;
  }

  bool _fillRemaining(int i, int j) {
    if ((j >= 9) && (i < (9 - 1))) {
      i = i + 1;
      j = 0;
    }
    if ((i >= 9) && (j >= 9)) {
      return true;
    }
    if (i < 3) {
      if (j < 3) {
        j = 3;
      }
    } else {
      if (i < (9 - 3)) {
        if (j == ((i ~/ 3) * 3)) {
          j = j + 3;
        }
      } else {
        if (j == (9 - 3)) {
          i = i + 1;
          j = 0;
          if (i >= 9) {
            return true;
          }
        }
      }
    }
    for (var number = 1; number <= 9; number++) {
      if (_checkIfSafe(i, j, number)) {
        _sudoku[i][j] = number;
        if (_fillRemaining(i, j + 1)) {
          return true;
        }
        _sudoku[i][j] = 0;
      }
    }
    return false;
  }

  void _removeClues() {
    while (_emptySquares > 0) {
      var row = _randomGenerator() - 1;
      var column = _randomGenerator() - 1;
      while (_sudoku[row][column] == 0) {
        row = _randomGenerator() - 1;
        column = _randomGenerator() - 1;
      }
      if (_uniqueSolution) {
        var backup = _sudoku[row][column];
        _sudoku[row][column] = 0;
        if (SudokuUtilities.hasUniqueSolution(_sudoku)) {
          _emptySquares--;
        } else {
          _sudoku[row][column] = backup;
        }
      } else {
        _sudoku[row][column] = 0;
        _emptySquares--;
      }
    }
  }
}


sudoku_solver_generator/lib/src/sudoku_solver_generator_base.dart
///
class SudokuSolver {
  late List<List<int>> _solvedGrid;
  int _counter = 0;

  SudokuSolver(List<List<int?>> grid) {
    // TODO: Allow accepting 1D Lists
    _solvedGrid = _replaceNullWithZero(grid);
    _solve(_solvedGrid);
  }

  List<int> get solution1D {
    var solvedGrid1D = List.generate(81, (i) => 0);
    var index = 0;
    for (var i = 0; i < 9; i++) {
      for (var j = 0; j < 9; j++) {
        solvedGrid1D[index] = _solvedGrid[i][j];
        index++;
      }
    }
    return solvedGrid1D;
  }

  List<List<int>> get solution2D {
    return _solvedGrid;
  }

  int get noOfSolutions {
    return _counter;
  }

  List<List<int>> _replaceNullWithZero(List<List<int?>> grid) {
    var notNullGrid = List.generate(9, (i) => List.generate(9, (j) => 0));
    for (var i = 0; i < 9; i++) {
      for (var j = 0; j < 9; j++) {
        notNullGrid[i][j] = grid[i][j] ?? 0;
      }
    }
    return notNullGrid;
  }

  List<int>? _findEmpty(List<List<int>> grid) {
    for (var i = 0; i < 9; i++) {
      for (var j = 0; j < 9; j++) {
        if (grid[i][j] == 0) {
          return [i, j];
        }
      }
    }
    return null;
  }

  bool _valid(List<List<int>> grid, int num, List<int> pos) {
    for (var i = 0; i < 9; i++) {
      if (grid[pos[0]][i] == num && pos[1] != i) {
        return false;
      }
    }
    for (var i = 0; i < 9; i++) {
      if (grid[i][pos[1]] == num && pos[0] != i) {
        return false;
      }
    }
    var boxX = pos[1] ~/ 3;
    var boxY = pos[0] ~/ 3;
    for (var i = boxY * 3; i < boxY * 3 + 3; i++) {
      for (var j = boxX * 3; j < boxX * 3 + 3; j++) {
        if (grid[i][j] == num && [i, j] != pos) {
          return false;
        }
      }
    }
    return true;
  }

  bool _solve(List<List<int>> grid) {
    int row;
    int col;
    if (_findEmpty(grid) == null) {
      _counter++;
      return true;
    } else {
      row = _findEmpty(grid)![0];
      col = _findEmpty(grid)![1];
    }
    for (var i = 1; i < 10; i++) {
      if (_valid(grid, i, [row, col])) {
        grid[row][col] = i;
        if (_solve(grid)) {
          return true;
        }
        grid[row][col] = 0;
      }
    }
    return false;
  }
}


sudoku_solver_generator/example/sudoku_solver_generator_example.dart
import 'package:sudoku_solver_generator/sudoku_solver_generator.dart';

void main() {
  // Solve Example
  var sudoku = <List<int?>>[
    [2, 1, 7, 4, 3, 5, 9, null, 8],
    [5, 8, 9, 0, 6, 2, 4, 3, null],
    [3, 4, 6, 0, 7, 9, 2, 5, 1],
    [1, 6, 2, 3, 0, 0, 5, 4, 0],
    [9, 0, 5, 2, 0, 4, 7, 8, 6],
    [4, 7, 8, 5, 9, 6, 1, 0, 0],
    [8, 0, 3, 0, 4, 1, 6, 7, 2],
    [0, 9, 4, 7, 2, 3, 8, 1, 5],
    [7, 2, 1, 0, 5, 0, 0, 0, 4],
  ];
  var sudokuSolved = SudokuSolver.solve(sudoku);
  SudokuUtilities.printSudoku(sudokuSolved);
  print(''); // ignore: avoid_print

  // Generate Example
  var sudokuGenerator = SudokuGenerator(emptySquares: 54);
  SudokuUtilities.printSudoku(sudokuGenerator.newSudoku);
  print(''); // ignore: avoid_print
  SudokuUtilities.printSudoku(sudokuGenerator.newSudokuSolved);
}


sudoku_solver_generator/lib/src/sudoku_exceptions_base.dart
/// Thrown when the configuration of the Sudoku is not valid.
///
/// Configuration of the Sudoku follows the rules of Sudoku.
///
/// Conditions:
/// * [List] of 9 [List]s of 9 [int]s
/// * Every row must have the numbers 1-9 without repetition
/// * Every column must have the numbers 1-9 without repetition
/// * Every 3x3 box must have the number 1-9 without repetition
/// * Only positions of non-zero and non-null values are considered
/// * Solvability of the puzzle is not considered
///
/// [SudokuUtilities.isValidConfiguration()] can be used to check for validity
/// to prevent this Exception.
class InvalidSudokuConfigurationException implements Exception {
  /// Returns the error message
  String errorMessage() =>
      'The configuration for the Sudoku puzzle is invalid.';

  @override
  String toString() => 'InvalidSudokuConfigurationException';
}

/// Thrown when the value of the [emptySquares] parameter of [SudokuGenerator]
/// is invalid.
///
/// Valid values: Any integer from 1-81 (inclusive).
class InvalidEmptySquaresException implements Exception {
  /// Returns the error message
  String errorMessage() =>
      'The number of empty squares cannot be less than 1 or more than 81.';

  @override
  String toString() => 'InvalidEmptySquaresException';
}

/// Thrown when the value of the [emptySquares] parameter of [SudokuGenerator]
/// is `> 54` and [uniqueSolution] is set to `true`.
///
/// This is because a unique solution is unlikely if there are less than
/// `27` clues.
class UnlikelyUniqueSolutionException implements Exception {
  /// Returns the error message
  String errorMessage() => 'The number of empty squares cannot more than 54 if '
      'a unique solution is needed.';

  @override
  String toString() => 'UnlikelyUniqueSolutionException';
}
